### PR TITLE
ovl_Shot_Sun Decompiled

### PR DESCRIPTION
**Pull request by sonicdcer**

### CODE REVIEW

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.h [hunk 1]
```diff
@@ -3,15 +3,20 @@
...
-    /* 0x0190 */ ShotSunActionFunc actionFunc;
-    /* 0x0194 */ char unk_194[0xC];
+    /* 0x000 */ Actor actor;
+    /* 0x144 */ ColliderCylinder collider;
+    /* 0x190 */ ShotSunActionFunc actionFunc;
+    /* 0x194 */ Vec3s hitboxPos;
+    /* 0x19A */ s16 timer;
+    /* 0x19C */ s8 unused;
+    /* 0x19D */ UNK_TYPE1 pad19D[3];
 } ShotSun; // size = 0x1A0
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.h:20]: This pad seems to be just struct padding, you can remove it.
```suggestion
    /* 0x19C */ s8 unused;
} ShotSun; // size = 0x1A0
```

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 2]
```diff
@@ -32,27 +30,155 @@ const ActorInit Shot_Sun_InitVars = {
...
+    s32 pad;
+    ShotSun* this = THIS;
+
+    if ((SHOTSUN_GET_FF(thisx) == 0x40) || (SHOTSUN_GET_FF(thisx) == 0x41)) {
+        this->unused = 0;
+        this->actor.flags |= ACTOR_FLAG_10;
+        this->actor.flags |= ACTOR_FLAG_2000000;
+        this->actionFunc = func_809738D0;
+        this->actor.flags |= ACTOR_FLAG_8000000;
+        ;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:63]: Is the extra semicolon required to match?

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 3]
```diff
@@ -32,27 +30,155 @@ const ActorInit Shot_Sun_InitVars = {
...
+                collectible->actor.speedXZ = 0.0f;
+            }
+        }
+
+        Actor_MarkForDeath(&this->actor);
+    } else if (!(this->actor.xzDistToPlayer > 120.0f)) {
+        if ((gSaveContext.save.time >= CLOCK_TIME(6, 30)) && (gSaveContext.save.time < CLOCK_TIME(7, 30))) {
+            cylinderPos.x = player->bodyPartsPos[7].x + (play->envCtx.unk_4 * 0.16666667f);
+            cylinderPos.y = (player->bodyPartsPos[7].y - 30.0f) + (play->envCtx.unk_8 * 0.16666667f);
+            cylinderPos.z = player->bodyPartsPos[7].z + (play->envCtx.unk_C * 0.16666667f);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:170]: Would those match with `/ 6.0f` or `* (1.0f / 6.0f)`?
```suggestion
            cylinderPos.x = player->bodyPartsPos[7].x + (play->envCtx.unk_4 / 6.0f);
            cylinderPos.y = (player->bodyPartsPos[7].y - 30.0f) + (play->envCtx.unk_8 / 6.0f);
            cylinderPos.z = player->bodyPartsPos[7].z + (play->envCtx.unk_C / 6.0f);
```

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 4]
```diff
@@ -32,27 +30,155 @@ const ActorInit Shot_Sun_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Shot_Sun/func_809738D0.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Shot_Sun/func_80973960.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Shot_Sun/ShotSun_Update.s")
+void ShotSun_Init(Actor* thisx, PlayState* play) {
+    s32 pad;
+    ShotSun* this = THIS;
+
+    if ((SHOTSUN_GET_FF(thisx) == 0x40) || (SHOTSUN_GET_FF(thisx) == 0x41)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:57]: Could you make defines for those values?

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 5]
```diff
@@ -32,27 +30,155 @@ const ActorInit Shot_Sun_InitVars = {
...
+    } else {
+        ActorCutscene_SetIntentToPlay(this->actor.cutscene);
+    }
+}
+
+void func_809738D0(ShotSun* this, PlayState* play) {
+    s32 params = SHOTSUN_GET_FF(&this->actor);
+
+    if (play->msgCtx.ocarinaMode == 3) {
+        switch (play->msgCtx.unk1202E) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:122]: Comparing with OoT it seems like `unk1202E` is `lastPlayedSong`
```suggestion
        switch (play->msgCtx.lastPlayedSong) {
```

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 6]
```diff
@@ -32,27 +30,155 @@ const ActorInit Shot_Sun_InitVars = {
...
+            case 10:
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:123]:
```suggestion
            case OCARINA_SONG_STORMS:
```

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 7]
```diff
@@ -32,27 +30,155 @@ const ActorInit Shot_Sun_InitVars = {
...
+                if (params == 0x41) {
+                    this->actionFunc = ShotSun_TriggerFairy;
+                    play->msgCtx.ocarinaMode = 4;
+                }
+                break;
+            case 11:
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:129]:
```suggestion
            case OCARINA_SONG_SUNS:
```

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c (hunk 3 above)
Comment (sonicdcer) [reply src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:170]: * (1.0f / 6.0f) worked

File: src/overlays/actors/ovl_En_Gs/z_en_gs.c [hunk 8]
```diff
@@ -275,7 +275,7 @@ void func_80998040(EnGs* this, PlayState* play) {
 void func_8099807C(EnGs* this, PlayState* play) {
     switch (play->msgCtx.ocarinaMode) {
         case 3:
-            switch (play->msgCtx.unk1202E) {
+            switch (play->msgCtx.lastPlayedSong) {
                 case 7:
                 case 8:
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Gs/z_en_gs.c:280]: Looks like the cases for this switch should also use the `OCARINA_SONG` enum.

**Review by hensldm — COMMENTED** [review]
1 thing

**Review by hensldm — APPROVED** [review]

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 9]
```diff
@@ -1,10 +1,11 @@
 /*
  * File: z_shot_sun.c
  * Overlay: ovl_Shot_Sun
- * Description:
+ * Description: Leftover sun hitbox from OOT
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:4]: Other types can also spawn fairies, which it looks like would still function, so worth mentioning that too.

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 10]
```diff
@@ -32,27 +30,154 @@ const ActorInit Shot_Sun_InitVars = {
...
+    if (play->msgCtx.ocarinaMode == 3) {
+        switch (play->msgCtx.lastPlayedSong) {
+            case OCARINA_SONG_STORMS:
+                if (params == SHOTSUN_41) {
+                    this->actionFunc = ShotSun_TriggerFairy;
+                    play->msgCtx.ocarinaMode = 4;
+                }
+                break;
+            case OCARINA_SONG_SUNS:
+                if (params == SHOTSUN_40) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:129]: Indeed, you can name the params based on the song required, `SHOTSUN_FAIRY_SPAWNER_STORMS` and `SHOTSUN_FAIRY_SPAWNER_SUNS`.

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 11]
```diff
@@ -32,27 +30,154 @@ const ActorInit Shot_Sun_InitVars = {
...
+        ActorCutscene_Stop(this->actor.cutscene);
+        switch (params) {
+            case SHOTSUN_40:
+                fairyType = 7;
+                break;
+            case SHOTSUN_41:
+                fairyType = 7;
+                break;
+        }
+        Actor_Spawn(&play->actorCtx, play, ACTOR_EN_ELF, this->actor.home.pos.x, this->actor.home.pos.y,
```
[the remark this answers is not in the document]
Comment (sonicdcer) [reply src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:96]: ???
s32 fairyType = 0; right there

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 12]
```diff
@@ -32,27 +30,156 @@ const ActorInit Shot_Sun_InitVars = {
...
+        switch (params) {
+            case SHOTSUN_FAIRY_SPAWNER_SUNS:
+                fairyType = 7;
+                break;
+            case SHOTSUN_FAIRY_SPAWNER_STORMS:
+                fairyType = 7;
+                break;
+        }
+        Actor_Spawn(&play->actorCtx, play, ACTOR_EN_ELF, this->actor.home.pos.x, this->actor.home.pos.y,
+                    this->actor.home.pos.z, 0, 0, 0, fairyType);
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:99]: We should be using the en_elf actor enum for fairyType

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c [hunk 13]
```diff
@@ -32,27 +30,156 @@ const ActorInit Shot_Sun_InitVars = {
...
+        this->timer = 50;
+        Actor_Spawn(&play->actorCtx, play, ACTOR_DEMO_KANKYO, this->actor.home.pos.x, this->actor.home.pos.y,
+                    this->actor.home.pos.z, 0, 0, 0, 0x11);
+        Audio_PlaySfxAtPos(&this->actor.projectedPos, NA_SE_EV_TRE_BOX_APPEAR);
+    } else {
+        ActorCutscene_SetIntentToPlay(this->actor.cutscene);
+    }
+}
+
+void func_809738D0(ShotSun* this, PlayState* play) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:119]: When is shot sun have a params value indicating it's used for an ocarina song? We can probably fairly easily name this something like UpdateForOcarina

File: src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c (hunk 12 above)
Comment (hensldm) [reply src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c:99]: Could just create a generic (something like `ENELF_TYPE_7`) enum in `z_en_elf.h`, that gets name/documented with `EnElf`

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**